### PR TITLE
Refactor /examples/pytorch/NanoGPT/*_train.py

### PR DESCRIPTION
**Pull request by Filtee**
### What changes were proposed in this pull request?

I refactored example/pytorch/NanoGPT/*_train.py, i.e. train.py, ds_train.py, and fsdp_train.py.

### Why are the changes needed?

These three files have almost identical code, with the main difference being the technology they use: fsdp, deepspeed, or DDP, which would just lead to a tiny little difference in code.

So I just decided to refactor all of them anyway. It seems that all of them works fine.

### Does this PR introduce any user-facing change?

Nope. It just relates to the example showing, not the main part of DLRover.

### How was this patch tested?

It works fine…

### CODE REVIEW

File: examples/pytorch/nanogpt/train.py [hunk 1]
```diff
@@ -328,16 +309,59 @@ def wrapper(*args, **kwargs):
...
+    """
+    model = model_params["model"]
+    optimizer = model_params["optimizer"]
+    train_loader = model_params["train_loader"]
+    checkpointer = ckpt_params["checkpointer"]
+    checkpoint_dir = ckpt_params["checkpoint_dir"]
+    os.makedirs(checkpoint_dir, exist_ok=True)
+
+    if ckpt_params["use_native"]:
+        ckpt_dict = torch.load(os.path.join(checkpoint_dir, "50.pt"))
```
Comment (workingloong) [inline examples/pytorch/nanogpt/train.py:346]: We should select a checkpoint file from the checkpoint directory not use a fixed`50.pt`.

Comment (Filtee) [reply examples/pytorch/nanogpt/train.py:346]: Yep, I also find this line of code a bit weird, but it's how the original source code was before my refactoring. 😯

(The following code comes from the original code.)

截屏2024-05-10 14 56 53

Should I make some modifications to it?

Comment (Filtee) [reply examples/pytorch/nanogpt/train.py:346]: Alright, I fixed it anyway.

**Review by workingloong — APPROVED** [review]
LGTM

**Pull request merged**